File: .cursor/tools/delete_stage_files.mjs
import fs from "node:fs";
import path from "node:path";
import readline from "node:readline";

const root = path.resolve(path.dirname(new URL(import.meta.url).pathname), "..", "..").replace(/^\/([A-Za-z]:)/, "$1");
const whitelistPath = path.join(root, ".cursor", "config", "stage-delete-whitelist.txt");
const trashRoot = path.join(root, ".trash", "soft-delete", "stage-files");

function readWhitelist(filePath) {
  return fs
    .readFileSync(filePath, "utf8")
    .split(/\r?\n/)
    .map((s) => s.trim())
    .filter((s) => s && !s.startsWith("#"));
}

function isStageFile(filePath) {
  if (path.basename(filePath).startsWith("STAGE_")) return true;
  const head = fs.readFileSync(filePath, "utf8").split(/\r?\n/).slice(0, 30).join("\n");
  return head.includes("Lifecycle: 阶段");
}

function collectCandidates() {
  const out = [];
  for (const rel of readWhitelist(whitelistPath)) {
    const base = path.join(root, rel);
    if (!fs.existsSync(base) || !fs.statSync(base).isDirectory()) continue;
    const stack = [base];
    while (stack.length) {
      const dir = stack.pop();
      for (const name of fs.readdirSync(dir)) {
        const full = path.join(dir, name);
        const st = fs.statSync(full);
        if (st.isDirectory()) stack.push(full);
        else if (name.endsWith(".md") && isStageFile(full)) out.push(full);
      }
    }
  }
  return out;
}

function softDelete(src) {
  const ext = path.extname(src);
  const stem = path.basename(src, ext);
  let target = path.join(trashRoot, path.basename(src));
  if (fs.existsSync(target)) {
    const stamp = new Date().toISOString().replace(/[-:TZ.]/g, "").slice(0, 14);
    target = path.join(trashRoot, `${stem}_${stamp}${ext}`);
  }
  fs.mkdirSync(path.dirname(target), { recursive: true });
  fs.renameSync(src, target);
  return target;
}

function ask(question, rl) {
  return new Promise((resolve) => rl.question(question, resolve));
}

if (!fs.existsSync(whitelistPath)) {
  console.error("Whitelist file missing:", whitelistPath);
  process.exit(1);
}

const candidates = collectCandidates();
if (!candidates.length) {
  console.log("No stage files found in whitelist paths.");
  process.exit(0);
}

console.log("Stage-file candidates:");
candidates.forEach((c, idx) => console.log(`${String(idx + 1).padStart(3, " ")}. ${path.relative(root, c)}`));

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
const yes = await ask("Type YES to continue to final confirmation: ", rl);
if (yes.trim() !== "YES") {
  console.log("Cancelled.");
  rl.close();
  process.exit(0);
}
const confirm = await ask("Type CONFIRM to soft-delete the above files: ", rl);
rl.close();
if (confirm.trim() !== "CONFIRM") {
  console.log("Cancelled.");
  process.exit(0);
}

const moved = candidates.map((src) => ({ src, dst: softDelete(src) }));
console.log(`Moved count: ${moved.length}`);
moved.forEach(({ src, dst }) => {
  console.log(`- ${path.relative(root, src)} -> ${path.relative(root, dst)}`);
});

File: .cursor/tools/delete_stage_files.py
#!/usr/bin/env python3
from __future__ import annotations

import shutil
from datetime import datetime
from pathlib import Path


def read_whitelist(path: Path) -> list[Path]:
    lines = []
    for raw in path.read_text(encoding="utf-8").splitlines():
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        lines.append(line)
    return [Path(x) for x in lines]


def is_stage_file(file_path: Path) -> bool:
    if file_path.name.startswith("STAGE_"):
        return True
    text = file_path.read_text(encoding="utf-8", errors="ignore")
    head = "\n".join(text.splitlines()[:30])
    return "Lifecycle: 阶段" in head


def soft_delete(src: Path, trash_root: Path) -> Path:
    target = trash_root / src.name
    if target.exists():
        stamp = datetime.now().strftime("%Y%m%d%H%M%S")
        target = trash_root / f"{src.stem}_{stamp}{src.suffix}"
    target.parent.mkdir(parents=True, exist_ok=True)
    shutil.move(str(src), str(target))
    return target


def main() -> int:
    root = Path(__file__).resolve().parents[2]
    wl_file = root / ".cursor" / "config" / "stage-delete-whitelist.txt"
    trash_root = root / ".trash" / "soft-delete" / "stage-files"
    if not wl_file.exists():
        print("Whitelist file missing:", wl_file)
        return 1

    candidates: list[Path] = []
    for rel in read_whitelist(wl_file):
        base = root / rel
        if not base.exists() or not base.is_dir():
            continue
        for file_path in base.rglob("*.md"):
            if is_stage_file(file_path):
                candidates.append(file_path)

    if not candidates:
        print("No stage files found in whitelist paths.")
        return 0

    print("Stage-file candidates:")
    for i, p in enumerate(candidates, start=1):
        print(f"{i:>3}. {p.relative_to(root)}")

    s1 = input("Type YES to continue to final confirmation: ").strip()
    if s1 != "YES":
        print("Cancelled.")
        return 0
    s2 = input("Type CONFIRM to soft-delete the above files: ").strip()
    if s2 != "CONFIRM":
        print("Cancelled.")
        return 0

    moved = []
    for p in candidates:
        moved.append((p, soft_delete(p, trash_root)))
    print(f"Moved count: {len(moved)}")
    for src, dst in moved:
        print(f"- {src.relative_to(root)} -> {dst.relative_to(root)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
